Link subtitle editor to timeline

bigfoodK opened the pull request:
![image](https://user-images.githubusercontent.com/38313680/140740956-0316b190-522f-40fb-9628-7b746e67bc14.png)


### Features
- Display subtitle editor on subtitle clip selected
- Constrain subtitle clip placement
- Edit subtitle clip placement on timeline

#### Subtitle clip constraint
- If previous clip exist, Clip's `startMs` must be greater than previous clip's `startMs + 300`

bigfoodK commented on luda/editor/client/src/timeline/cameraTrack/CameraTrackBody.ts:
```diff
@@ -110,6 +110,8 @@ function pushClipsForward(state: TimelineState, clips: Clip[]) {
     const { clipId } = state.actionState;
     const draggingClip = clips.find((clip) => clip.id === clipId);
     if (!draggingClip) {
+      // TODO: It may be a clip of another track.
+      return;
       throw new Error("clip not found");
     }
```
If dragging clip is other track's clip. It will be undefined

bigfoodK replied:
Also here
https://github.com/NamseEnt/namseent/pull/18/files#diff-9afdf27459f954adc859a72610390222c897bd4840f471599a53ea8c1157bad3R174-R175

bigfoodK commented on luda/editor/client/src/timeline/clip/ClipComponent.ts:
```diff
@@ -51,6 +53,17 @@ export const ClipComponent: Render<
     height: height - 2,
   };
 
+  // TODO: If possible, it should be integrated with ClipComponent.
+  if (isSubtitleClip(clip)) {
+    return SubtitleClipComponent(
+      {
+        clip,
+        timelineState,
+      },
+      props,
+    );
+  }
+
```
There is two `clipComponent`. `SubtitleClipComponent` and `ClipComponent`.
It is because `SubtitleClipComponent` is composed of several pathes rather than on rect.
Should we combine them into one `clipComponent`?

bigfoodK replied:
I added function `renderClip`.
looks https://github.com/NamseEnt/namseent/commit/0763f1f88e6c378eff9e595c69894729119f1903#diff-fc1360505b3444f28e11154cf00638192f1d9987b7c4b136c4de8467175518ed
with https://github.com/NamseEnt/namseent/pull/18/commits/0763f1f88e6c378eff9e595c69894729119f1903

namse commented:
@bigfoodK 
[QA]
Progress
1. Click subtitle so i select it, it changes the color.
2. When i press track body empty space, 

Expected
- Change subtitle clip color back as normal
- Close subtitle editor

What actually happened
- No color changed, still selected color.
- It keep open subtitle editor.

bigfoodK commented:
> @bigfoodK [QA] Progress
> 
> 1. Click subtitle so i select it, it changes the color.
> 2. When i press track body empty space,
> 
> Expected
> 
> * Change subtitle clip color back as normal
> * Close subtitle editor
> 
> What actually happened
> 
> * No color changed, still selected color.
> * It keep open subtitle editor.

I'll fix it now. :)

namse commented:
@bigfoodK
[QA]
## Progress
1. put multiple subtitles in subtitle track.
2. select subtitle and drag it to move from left to right

## Expected
- selected subtitle clip just move right and right
- If there is other clip, and if my mouse is on that clip's head area, it block my movement.
- But if i pass over it, my clip jump and move right.

## What actually happened
- Selected subtitle clip just push the right clip.
- Then, suddenly jumped.

bigfoodK commented:
> @bigfoodK [QA]
> 
> ## Progress
> 1. put multiple subtitles in subtitle track.
> 2. select subtitle and drag it to move from left to right
> 
> ## Expected
> * selected subtitle clip just move right and right
> * If there is other clip, and if my mouse is on that clip's head area, it block my movement.
> * But if i pass over it, my clip jump and move right.
> 
> ## What actually happened
> * Selected subtitle clip just push the right clip.
> * Then, suddenly jumped.

You mean the other clips shouldn't move, right?

namse commented:
> > @bigfoodK [QA]
> > ## Progress
> > 
> > 1. put multiple subtitles in subtitle track.
> > 2. select subtitle and drag it to move from left to right
> > 
> > ## Expected
> > 
> > * selected subtitle clip just move right and right
> > * If there is other clip, and if my mouse is on that clip's head area, it block my movement.
> > * But if i pass over it, my clip jump and move right.
> > 
> > ## What actually happened
> > 
> > * Selected subtitle clip just push the right clip.
> > * Then, suddenly jumped.
> 
> @namse You mean the other clips shouldn't move, While moving the clip. right?

Yes because I grab the clip to move it, not other thing.

bigfoodK commented on luda/editor/client/src/index.ts:
```diff
@@ -114,10 +153,10 @@ startEngine(
+          y: margin,
           width: 400,
-          height: 800,
+          height: subtitleEditorHeight,
```
All editor uses same layout
With https://github.com/NamseEnt/namseent/pull/18/commits/86679c81f27ce7dd700fed189b3ef2fa807e2b73

bigfoodK commented on luda/editor/client/src/util/color/convertColorToHsl.ts:
```diff
@@ -0,0 +1,42 @@
+export function convertColorToHsl(color: Float32Array) {
```
This function moved to engine's `Convert`
with https://github.com/NamseEnt/namseent/pull/18/commits/07adc2f56ad96628a677ce3f960436792a87df5b

bigfoodK commented on luda/editor/client/src/subtitleEditor/renderSubtitleEditor.ts:
```diff
@@ -1,31 +1,44 @@
+  {},
+  {
+    subtitleEditor: SubtitleEditorState;
+    timeline: TimelineState;
```
Changed props and state
with https://github.com/NamseEnt/namseent/pull/18/commits/c7103797ece2d0b5a0d549fcee06dd31fd875689

bigfoodK commented on luda/editor/client/src/timeline/operations/createClip.ts:
```diff
@@ -36,6 +37,40 @@ export function createClip({
+        subtitle: {
+          text: "Input text here",
+          fontType: {
+            fontWeight: 400,
```
I agree. Fixed it
with https://github.com/NamseEnt/namseent/pull/18/commits/d495dde9505560699e836bd886c72cfd7ddfde75

bigfoodK commented on luda/editor/client/src/timeline/renderTimelineBody.ts:
```diff
@@ -97,11 +97,16 @@ const TrackBody: Render<
+      return DefaultTrackBody(
+        {
+          timelineState: timelineState,
+          track: track,
```
Removed it. thx :D
with https://github.com/NamseEnt/namseent/pull/18/commits/4122dba1700d744d2600d98f454ec01b68645fa7

bigfoodK commented on luda/editor/client/src/util/clamp.ts:
```diff
@@ -0,0 +1,3 @@
+export function clamp(value: number, min: number = 0, max: number = 1) {
```
This function moved engine's Mathu
with https://github.com/NamseEnt/namseent/pull/18/commits/07adc2f56ad96628a677ce3f960436792a87df5b

bigfoodK commented on luda/editor/client/src/util/color/brighterColor.ts:
```diff
@@ -0,0 +1,14 @@
+import { clamp } from "../clamp";
+import { convertColorToHsl } from "./convertColorToHsl";
+
+export function brighterColor01(color: Float32Array, amount: number) {
```
This function moved into engine's `CololUtil`
with https://github.com/NamseEnt/namseent/pull/18/commits/07adc2f56ad96628a677ce3f960436792a87df5b

bigfoodK commented on luda/editor/client/src/timeline/DefaultTrackBody.ts:
```diff
@@ -32,22 +34,151 @@ export const DefaultTrackBody: Render<
+    clipComponentProps,
+  } = props;
+  if (!noEmit) {
+    console.log(123);
```
Fixed it
with https://github.com/NamseEnt/namseent/pull/18/commits/996340e0def18dd72136209d35e57ab311b61ab9

bigfoodK commented on luda/editor/client/src/timeline/DefaultTrackBody.ts:
```diff
@@ -32,22 +34,151 @@ export const DefaultTrackBody: Render<
   ];
 };
+
+function renderClipWithPlacementConstraint(props: {
```
Fixed it
with https://github.com/NamseEnt/namseent/pull/18/commits/996340e0def18dd72136209d35e57ab311b61ab9

bigfoodK commented on luda/editor/client/src/timeline/clip/SubtitleClipComponent.ts:
```diff
@@ -0,0 +1,261 @@
+        }
+
+        engine.mouseEvent.onMouseDown((mouseEvent) => {
+          if (mouseIn) {
```
I use `fakeDraggingClip` concept
with https://github.com/NamseEnt/namseent/pull/18/commits/996340e0def18dd72136209d35e57ab311b61ab9

And fixed about mouse event
with https://github.com/NamseEnt/namseent/pull/18/commits/01fdc51d52f0c8fea8d1d879c926e812a1352c9c

namse replied:
I super recommend, removing `engine.mouseEvent.onMouseDown((mouseEvent) => {` from AfterDraw and just put this function into renderingData above.  There is head renderingData and tail renderingData.
Before, we share `mouseIn` with outside of event handler, it looks better, but now, we re-calculate `mouseIn`, then I believe there are no reason to put this function in here..
One of big benefit if you move this function in renderingData's OnClick is, you can make it sure that mouse in tail or head because that rendering data are for tail or head.

namse commented:
@bigfoodK
[QA]
## Progress
1. put multiple subtitles clips in subtitle track.
2. select subtitle clip and move it over other clip

## Expected
- the clip I selected jump over right side because every clip in subtitle track head should keep gap.

## What actually happened
- the clip i selected and moved just stop on other clip.

https://user-images.githubusercontent.com/3580430/140915096-673c14c0-7152-4c7a-9013-fd1fd1f531be.mp4

namse commented on luda/editor/client/src/timeline/clip/renderClip.ts:
```diff
@@ -0,0 +1,42 @@
+  const { clip } = state;
+
+  if (isCameraClip(clip)) {
+    console.log("camera");
```
oupppss

bigfoodK replied:
I removed it. thx ;)
with https://github.com/NamseEnt/namseent/pull/18/commits/cfdc376d5fbe7f729582699affe87cb6a8c52e5f

namse commented on luda/editor/client/src/timeline/clip/renderClip.ts:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  throw new Error(
+    "No rendering implementation for this clip type. Please implement it.",
```
I think it is ok because we only have 2 track, but I recommend returning `ClipComponent`. with sashComponent `luda/editor/client/src/timeline/clip/Sash.ts`.

bigfoodK replied:
That looks better. Fixed
with https://github.com/NamseEnt/namseent/pull/18/commits/1b73b02f3bdf4f1bba8a580d29d8062d1ddc4457

namse commented on luda/editor/client/src/timeline/DefaultTrackBody.ts:
```diff
@@ -1,17 +1,34 @@
 > = (state, props) => {
-  const { clips } = props.track;
+  const { clips } = state.track;
+  let selectedClip: Clip | undefined = undefined;
```
I think this is unused variable. let's remove it

bigfoodK replied:
Yes. That's unused. Removed
with https://github.com/NamseEnt/namseent/pull/18/commits/7f1324abc6a5ad3addf6ded835c956a5d553411e

namse commented on luda/editor/client/src/timeline/DefaultTrackBody.ts:
```diff
@@ -32,22 +49,134 @@ export const DefaultTrackBody: Render<
         }
       },
     }),
+    Rect({
```
I think this and below Rect are not necessary for *Default*TrackBody. I recommend making SubtitleTrackBody if you want different looks.

bigfoodK replied:
I divided subtitle track body
with https://github.com/NamseEnt/namseent/pull/18/commits/bb1f6853c663a4bb668e56ab67b94365834a2c1e

namse commented on luda/editor/client/src/timeline/DefaultTrackBody.ts:
```diff
@@ -32,22 +49,134 @@ export const DefaultTrackBody: Render<
   ];
 };
+
+function constrainDraggingClipPlacement(state: TimelineState, clips: Clip[]) {
```
I think this logic is only for subtitle. Make this function without `isSubtitleClip` but with some arguments like the minimum gap, or just put this function in SubtitleTrackBody.

bigfoodK commented:
> @bigfoodK [QA]
> 
> ## Progress
> 1. put multiple subtitles clips in subtitle track.
> 2. select subtitle clip and move it over other clip
> 
> ## Expected
> * the clip I selected jump over right side because every clip in subtitle track head should keep gap.
> 
> ## What actually happened
> * the clip i selected and moved just stop on other clip.
> 
>  _2021_11_09_20_20_09_625.mp4

https://github.com/NamseEnt/namseent/pull/18/commits/1510e1e2a1d7413e282bbce7835c25ead1668f9a
Subtitle clips **that are not dragging** should hold there place.
So I have to know which clip is being dragged.

But, when an `actionState` exists, the location of the clip is modified by `handleActionState`  in `timeline.afterDraw`.
And, when not exists. I can't find which clip is being dragged.

So I added `terminatePhase`. For, `actionState` exists but no `handleActionState` runs.
Is it ok to use it?

namse commented on luda/editor/client/src/index.ts:
```diff
@@ -25,6 +26,42 @@ startEngine(
           id: "track1",
+          type: TrackType.subtitle,
+          clips: [
+            {
```
I recommend using `createClip` for this if you not test something with custom data.

bigfoodK replied:
Fixed
with https://github.com/NamseEnt/namseent/pull/18/commits/27f61849cfa8272b4969baa72cee0d263caa4e7a

namse commented on luda/editor/client/src/type.ts:
```diff
@@ -23,3 +23,40 @@ export type CameraAngle = {
+  fontWeight: FontWeight;
+};
+export type SubtitleStyle = {
+  color: Float32Array;
```
Let's export `Color` from `namui/src/index.ts` like Paint and use it.

bigfoodK replied:
Not CanvasKit.Color?

namse replied:
Check `namui/src/index.ts` 's Paint type, how it export Paint of canvaskit.

```
export { Paint } from "canvaskit-wasm";

bigfoodK replied:
Ah. Got it.
Reexported it from engin's index.ts
with https://github.com/NamseEnt/namseent/pull/18/commits/cbc870c876c447dfb4c4eee75151e5fcad4d88c3

namse commented:
> > @bigfoodK [QA]
> > ## Progress
> > 
> > 1. put multiple subtitles clips in subtitle track.
> > 2. select subtitle clip and move it over other clip
> > 
> > ## Expected
> > 
> > * the clip I selected jump over right side because every clip in subtitle track head should keep gap.
> > 
> > ## What actually happened
> > 
> > * the clip i selected and moved just stop on other clip.
> > 
> > _2021_11_09_20_20_09_625.mp4
> 
> [26933c0](https://github.com/NamseEnt/namseent/commit/26933c030d5a631208df6f54b47ccfcbf80556e5) Subtitle clips **that are not dragging** should hold there place. So I have to know which clip is being dragged.
> 
> But, when an `actionState` exists, the location of the clip is modified by `handleActionState` in `timeline.afterDraw`. And, when not exists. I can't find which clip is being dragged.
> 
> So I added `terminatePhase`. For, `actionState` exists but no `handleActionState` runs. Is it ok to use it?

actionState's type DragClip has clip Id so you can know which clip is dragging, don't you?

namse commented on luda/editor/client/src/timeline/renderSubtitleTrackBody.ts:
```diff
@@ -0,0 +1,178 @@
+        },
+      },
+    }),
+    Rect({
```
I think this Rect is duplicated with first Rect. Could you check it?

bigfoodK replied:
That was hack for redrawing outline. because while drawing middle line, I overwrited outline with middle line color.
Thanks to `BorderPosition.outside`, It has been replaced easily by using `Clip`
https://github.com/NamseEnt/namseent/pull/18/commits/78cea485cd547097168c79ae02e8014de04a2dac

bigfoodK commented:
> > > @bigfoodK [QA]
> > > ## Progress
> > > 
> > > 1. put multiple subtitles clips in subtitle track.
> > > 2. select subtitle clip and move it over other clip
> > > 
> > > ## Expected
> > > 
> > > * the clip I selected jump over right side because every clip in subtitle track head should keep gap.
> > > 
> > > ## What actually happened
> > > 
> > > * the clip i selected and moved just stop on other clip.
> > > 
> > > _2021_11_09_20_20_09_625.mp4
> > 
> > 
> > [26933c0](https://github.com/NamseEnt/namseent/commit/26933c030d5a631208df6f54b47ccfcbf80556e5) Subtitle clips **that are not dragging** should hold there place. So I have to know which clip is being dragged.
> > But, when an `actionState` exists, the location of the clip is modified by `handleActionState` in `timeline.afterDraw`. And, when not exists. I can't find which clip is being dragged.
> > So I added `terminatePhase`. For, `actionState` exists but no `handleActionState` runs. Is it ok to use it?
> 
> actionState's type DragClip has clip Id so you can know which clip is dragging, don't you?
> 
> (added comment) AH, did you know CameraTrackBody always update dragging clip's location in `pushClipsForward`?
> 
> Let me see. You can know the location where you should put the clip, right? You can put your clip where should it be in every frame if you are dragging that clip.
> 
> I have no idea why you need to know when user stops dragging.

### Why subtitleClip need to know dragging clip
`pushClipsForward` works without knowing which one is dragging clip. Because it simply list clips by index.
But in subtitle clips. other clips that is not dragging should be fixed. And only dragging clip can be moved.
So we have to know which one is dragging clip.
And only can do that when **`actionState` is exsist.**

### Why `constrainDraggingClipPlacement` didn't work before
`constrainDraggingClipPlacement` fixes clip's placement in **rendering time**.
But `handleActionState` fixes clip's placement again in afterDraw time.

As mentioned above, 'constrainDraggingClipPlacement' must be executed while the actionState is exsit.
But if actionState exsit, `handleActionState` will fix clip's placement to mouse position after 'constrainDraggingClipPlacement' executed.

So `actionState`, that has **which one is dragging clip but not execute `handleActionState`**, was needed.
In `terminated` phase. It skips `handleActionState`
https://github.com/NamseEnt/namseent/pull/18/files#diff-13db2fc619b0f67995dfe4bc655112d88cdcd8cb518c71e6771e279ade4c5861R125-R132

`know when user stops dragging` is not important

namse commented:
@bigfoodK Please review this commit. I re-implement dragging thing. I put comment inside of commit in github.
ca38a27899aa3550a5a0dd9681c53dad013a23bd

bigfoodK commented:
> @bigfoodK Please review this commit. I re-implement dragging thing. I put comment inside of commit in github. [ca38a27](https://github.com/NamseEnt/namseent/commit/ca38a27899aa3550a5a0dd9681c53dad013a23bd) If you feel ok, then use this commit please.

Looks really good. That's what I wanted
Use this method by https://github.com/NamseEnt/namseent/pull/18/commits/630fcff4652ea283e0ad6f56fec51e25425ae5a5

namse commented:
@bigfoodK You did a good job :)
